### ISSUE TITLE
Link using lld

### ISSUE THREAD

**Issue by phil-opp** [phil-opp/blog_os#370]
Use the cross-platform lld linker of the LLVM project. Compared to the binutils linker, no manual cross-compilation is required.

**Comment by gil0mendes** [comment]
@phil-opp One question. lld already has support for removing unused sections? The `gc-sections` of ld? That was one of the drawbacks that I found when used the lld.

**Comment by robert-w-gries** [comment]
@gil0mendes `ld.lld` supports the `gc-sections` option, at least in version `4.0.1`.  I actually ran into a different issue where `nmagic` was missing as an option in a newer release.

**Comment by robert-w-gries** [comment]
I had a chance to look into the issue again.  I managed to get latest `ld.lld` working for `blog_os`

### Notes regarding flags

* [`gc-sections`](https://github.com/llvm-project/llvm-project-20170507/commit/fa3c79239987ffb1297144fc5767e54987941988) is available in the 4.x and later.
* `nmagic` was removed at some point before `4.0` release.  I believe it was removed when they moved to the new ELF linker.
* [`omagic`](https://github.com/llvm-project/llvm-project-20170507/commit/b0641b2bc14d7ad9375a8ca978a43285ef3c375d) seems to be our replacement for `nmagic`. This flag differs from `nmagic` in that it sets the text and data sections to be readable and **writable**, in addition to disabling page-alignment. I'm not experienced enough with linkers to know whether this is allowed for our kernel.

### Changes to `linker.ld`

In addition to using the `omagic` flag, I had to add `ALIGN`s to `linker.ld` in order to align the pages properly
<details><summary>Click to expand diff</summary>

```diff
diff --git a/src/arch/x86_64/linker.ld b/src/arch/x86_64/linker.ld
index 62f8f26..d5d5acc 100644
--- a/src/arch/x86_64/linker.ld
+++ b/src/arch/x86_64/linker.ld
@@ -14,7 +14,7 @@ ENTRY(start)
 SECTIONS {
   . = 1M;
 
-  .rodata :
+  .rodata : ALIGN(4K)
   {
     /* ensure that the multiboot header is at the beginning */
     KEEP(*(.multiboot_header))
@@ -22,31 +22,31 @@ SECTIONS {
     . = ALIGN(4K);
   }
 
-  .text :
+  .text : ALIGN(4K)
   {
     *(.text .text.*)
     . = ALIGN(4K);
   }
 
-  .data :
+  .data : ALIGN(4K)
   {
     *(.data .data.*)
     . = ALIGN(4K);
   }
 
-  .bss :
+  .bss : ALIGN(4K)
   {
     *(.bss .bss.*)
     . = ALIGN(4K);
   }
 
-  .got :
+  .got : ALIGN(4K)
   {
     *(.got)
     . = ALIGN(4K);
   }
 
-  .got.plt :
+  .got.plt : ALIGN(4K)
```

</details>

### Testing for various linkers

* [x] `ld` - version `2.29-6.fc27` on my system
* [x] `ld.lld v6.0.0` - This was the latest code available on the [master branch](https://github.com/llvm-project/llvm-project-20170507)
* [ ] `ld.lld v4.0.1` - This was the binary available on Fedora

#### `ld.lld` version 4.0.1

I could not get this version to work with `blog_os`.  

```
PANIC in src/memory/paging/mod.rs at line 209:
    sections need to be page aligned
```

Dumping the sections in `kernel-x86_64.bin` reveals a duplicate `.rodata` section that is mis-aligned.

```
[rob@localhost blog_os]$ objdump -h build/kernel-x86_64.bin

build/kernel-x86_64.bin:     file format elf64-x86-64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .rodata       00002c25  0000000000100000  0000000000100000  00001000  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  1 .rodata       000003d8  0000000000102c28  0000000000102c28  00003c28  2**3
```

I believe there was an issue with alignment in `ld.lld 4.0.1` that was fixed later.  I am going to build `llvm 5.0.0` to narrow down where it was fixed.

**Comment by gil0mendes** [comment]
Hey, before all, great report! 👏

Now. I had some time to look at this, and you are right. Forget what I said before, my problem was exactly the same. The `-n` (or `-nmagic`) flag is missing, and not using this flag led to issues since sections' pages become misaligned.

**Comment by phil-opp** [comment]
@robert-w-gries Awesome work!

> omagic seems to be our replacement for nmagic. This flag differs from nmagic in that it sets the text and data sections to be readable and writable, in addition to disabling page-alignment. I'm not experienced enough with linkers to know whether this is allowed for our kernel.

Hmm, ideally we don't want a writable text section. But it's great that this seems to be the only issue with the latest lld!

>ld.lld version 4.0.1
> 
>I could not get this version to work with blog_os.

It seems like lld will be distributed with the compiler soon (see https://github.com/rust-lang/rust/issues/39915#issuecomment-346960292, which is currently blocked on the [upgrade to LLVM 5](https://github.com/rust-lang/rust/issues/43370)). So we might not need to support older lld versions.

**Comment by Evrey** [comment]
> Hmm, ideally we don't want a writable text section.

In fashion of the general security principle of *"never trust a damn thing"*, you can always set the NX bit for the pages mapping your code sections, no matter what the kernel loader actually did. *(And the read-only bit for `.rodata` and friends.)*

**Comment by steveklabnik** [comment]
This can now be closed, right?

**Comment by phil-opp** [comment]
Yes, closing in favor of #360. Thank you all for helping!
